### PR TITLE
DDP-4655 kit zip codes

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

* Added new `ZIP_CODE` rule type for kit configurations.
* Added study-builder support for specifying these new rules.
* Our kit check/creation code now checks the zip codes if configured.
* `POST /invitation-check` API now looks for these zip code rules and checks request.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [x] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [x] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/docs/specification/src/endpoints/studies.invitation-check.yml [hunk 1]
```diff
@@ -17,6 +17,7 @@ post:
             - auth0Domain
             - invitationId
             - recaptchaToken
+            - qualificationDetails
```
Comment (yufengwng) [inline pepper-apis/docs/specification/src/endpoints/studies.invitation-check.yml:19]: Screen Shot 2020-05-15 at 12 31 05 PM

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiKitRules.java [hunk 2]
```diff
@@ -39,16 +37,16 @@
...
+        long ruleId = insertRule(kitRuleType);
         switch (kitRuleType) {
             case COUNTRY:
                 insertCountryRule(ruleId, pexOrCountryId);
                 break;
             case PEX:
                 insertPexRule(ruleId, pexOrCountryId);
                 break;
+            case ZIP_CODE:
+                throw new DaoException("do not use for kit type " + kitRuleType);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiKitRules.java:49]: This should be okay. These are only used in some tests. I will refactor these DAOs later.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitConfigurationDao.java [hunk 3]
```diff
@@ -96,59 +114,65 @@ default long addCountryRule(long configId, String countryCode) {
...
+            + "  from kit_configuration__kit_rule as kckr"
+            + "  join kit_rule as kr on kr.kit_rule_id = kckr.kit_rule_id"
+            + "  join kit_rule_type as krt on krt.kit_rule_type_id = kr.kit_rule_type_id"
+            + "  left join kit_pex_rule as kp on kp.kit_rule_id = kr.kit_rule_id"
+            + "  left join expression as pe on pe.expression_id = kp.expression_id"
+            + "  left join kit_country_rule as kc on kc.kit_rule_id = kr.kit_rule_id"
+            + " where kckr.kit_configuration_id = :configId")
+    @UseRowReducer(KitRuleReducer.class)
+    List<KitRule> findRulesByConfigId(@Bind("configId") long configId);
+
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitConfigurationDao.java:152]: We're going to be querying these more often now, so did some minimal refactors to optimize the queries. 
There's more we can do later.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitConfigurationDao.java [hunk 4]
```diff
@@ -96,59 +114,65 @@ default long addCountryRule(long configId, String countryCode) {
...
+            KitRuleType type = KitRuleType.valueOf(view.getColumn("kit_rule_type", String.class));
+            KitRule rule;
+            switch (type) {
                 case PEX:
-                    KitPexRuleDto kitPexRuleDto = jdbiKitRules.getKitPexRuleById(kitRuleDto.getId()).get();
-                    PexInterpreter pexInterpreter = new TreeWalkInterpreter();
-                    String expression = jdbiExpression.getExpressionById(kitPexRuleDto.getExpressionId());
-                    kitRules.add(new KitPexRule(pexInterpreter, expression));
+                    String expression = view.getColumn("pex_expression", String.class);
+                    rule = new KitPexRule(ruleId, new TreeWalkInterpreter(), expression);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitConfigurationDao.java:162]: Don't like how we have to `new` up a new interpreter for each one, but refactoring will require a lot more changes than what I can afford now. Most kit configurations only have one of these kit pex rules, so should be fine for now.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/kit/KitConfiguration.java [hunk 5]
```diff
@@ -41,28 +44,44 @@ public String getStudyGuid() {
         return studyGuid;
     }
 
+    public Collection<KitRule> getRules() {
+        return List.copyOf(rules);
+    }
+
     public boolean evaluate(Handle handle, String userGuid) {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/kit/KitConfiguration.java:51]: This method is called as part of the kit check/creation process in Housekeeping. The new code here should do what we need -- that is, _not_ creating a new kit if zip code doesn't match.

File: study-builder/studies/testboston/study_kits.conf [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      "type": "SALIVA",
+      "quantity": 1,
+      "rules": [
+        {
+          "type": "COUNTRY",
+          "country": "us"     # United States
+        },
+        {
+          "type": "ZIP_CODE",
+          "zipCodes": [
```
Comment (yufengwng) [inline study-builder/studies/testboston/study_kits.conf:13]: Here's all the ones we need for `testboston`. See the attachment in DDP-4655.

File: pepper-apis/docs/specification/src/endpoints/studies.invitation-check.yml [hunk 7]
```diff
@@ -30,8 +31,23 @@ post:
             recaptchaToken:
               description: the user's reCAPTCHA token
               type: string
+            qualificationDetails:
+              description: additional user qualifications for checking the invitation
+              type: object
```
Comment (MocanaAtBroad) [inline pepper-apis/docs/specification/src/endpoints/studies.invitation-check.yml:32]: I am guessing you leaving the format of these object vague on purpose. Might be better to create schema for it and leave the fields  optional (like zipCode). We can adjust over time.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitConfigurationDao.java [hunk 8]
```diff
@@ -96,59 +114,65 @@ default long addCountryRule(long configId, String countryCode) {
...
+            + "       (select country_code from country where country_id = kc.country_id) as country_code,"
+            + "       (select group_concat(zip_code separator ',') from kit_zip_code where kit_rule_id = kr.kit_rule_id) as zip_codes"
+            + "  from kit_configuration__kit_rule as kckr"
+            + "  join kit_rule as kr on kr.kit_rule_id = kckr.kit_rule_id"
+            + "  join kit_rule_type as krt on krt.kit_rule_type_id = kr.kit_rule_type_id"
+            + "  left join kit_pex_rule as kp on kp.kit_rule_id = kr.kit_rule_id"
+            + "  left join expression as pe on pe.expression_id = kp.expression_id"
+            + "  left join kit_country_rule as kc on kc.kit_rule_id = kr.kit_rule_id"
+            + " where kckr.kit_configuration_id = :configId")
+    @UseRowReducer(KitRuleReducer.class)
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitConfigurationDao.java:150]: Could not figure it if you had test that made sure, at least indirectly, that this query works as expected. Do you?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/kit/KitZipCodeRule.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    public KitZipCodeRule(long id, Set<String> zipCodes) {
+        super(id, KitRuleType.ZIP_CODE);
+        this.zipCodes = new HashSet<>(zipCodes);
+    }
+
+    @Override
+    public boolean validate(Handle handle, String inputZipCode) {
+        return inputZipCode != null && zipCodes.contains(inputZipCode.trim());
+    }
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/kit/KitZipCodeRule.java:30]: You will have a problem here with a lot (most?) of the zip codes that we collect being of the extended form, e.g.: 
**02461-1314**
This is a US thing.  You already associating the zip codes with countries, so maybe this is a US type of match?

**Review by MocanaAtBroad — CHANGES_REQUESTED** [review]
Looks good overall, but will need to address extended zip codes.

File: pepper-apis/docs/specification/src/endpoints/studies.invitation-check.yml (hunk 7 above)
Comment (yufengwng) [reply pepper-apis/docs/specification/src/endpoints/studies.invitation-check.yml:32]: Will add.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitConfigurationDao.java (hunk 8 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/KitConfigurationDao.java:150]: The `testKitZipCodeRule()` test in `KitConfigurationTest.java` is exercising these `insert`s and `select`s.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/kit/KitZipCodeRule.java (hunk 9 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/model/kit/KitZipCodeRule.java:30]: Ah, good point! Based on a cursory reading, it sounds like the last 4 digits are specific to addresses and mail delivery, and not to an area/region. I'm assuming that we just want the area zip code, so potential fix here is to strip out that second part from the input before matching?

Comment (MocanaAtBroad) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/model/kit/KitZipCodeRule.java:30]: I think that would be fine. If we are confident in this place of the code that the user ZIPs have the proper `\d{5}(-\d4)?` format then you can just grab the first 5 chars and compare against our "supported zip code" list.

**Review by MocanaAtBroad — DISMISSED** [review]

**Review by MocanaAtBroad — COMMENTED** [review]
Let's get this baby merged!

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**